blivet/devices/lvm.py

```diff
@@ -1789,8 +1789,132 @@ def populate_ksdata(self, data):
         data.pool_name = self.pool.lvname
 
 
+class LVMVDOPoolMixin(object):
+    def __init__(self):
+        self._lvs = []
+
+    @property
+    def is_vdo_pool(self):
+        return self.seg_type == "vdo-pool"
+
+    @property
+    def type(self):
+        return "lvmvdopool"
+
+    @property
+    def resizable(self):
+        return False
+
+    @util.requires_property("is_vdo_pool")
+    def _add_log_vol(self, lv):
+        """ Add an LV to this VDO pool. """
+        if lv in self._lvs:
+            raise ValueError("lv is already part of this VDO pool")
+
+        self.vg._add_log_vol(lv)
+        log.debug("Adding %s/%s to %s", lv.name, lv.size, self.name)
+        self._lvs.append(lv)
+
+    @util.requires_property("is_vdo_pool")
+    def _remove_log_vol(self, lv):
+        """ Remove an LV from this VDO pool. """
+        if lv not in self._lvs:
+            raise ValueError("specified lv is not part of this VDO pool")
+
+        self._lvs.remove(lv)
+        self.vg._remove_log_vol(lv)
+
+    @property
+    @util.requires_property("is_vdo_pool")
+    def lvs(self):
+        """ A list of this VDO pool's LVs """
+        return self._lvs[:]     # we don't want folks changing our list
+
+    @property
+    def direct(self):
+        """ Is this device directly accessible? """
+        return False
+
+    def _create(self):
+        """ Create the device. """
+        raise NotImplementedError
+
+
+class LVMVDOLogicalVolumeMixin(object):
```

Comment:
How similar is this to `LVMThinLogicalVolumeMixin`? A little weird, but aren't they mostly the same?

Reply:
The functions and properties are similar but usually differ a little. And the mixins are already very complicated and I'd like to avoid having `LVMVDOLogicalVolumeMixin` inheriting from `LVMThinLogicalVolumeMixin` and `LVMLogicalVolumeDevice` depending on both. This could go wrong very fast and complex dependencies in different versions of Python are huge mess.

I was thinking about adding another class, something like `LVMThinVolumeBase` (and `LVMPoolBase` for thinpool and vdopool) but I think this would make the code even less readable than now.

And also these two will differ more when we add support for creating VDO LVs/pools.